### ISSUE TITLE
FreeBSD symlinks

### ISSUE THREAD

**Issue by ghost** [jimsalterjrs/sanoid#85]
The INSTALL file reads:
```
  Simplest path workaround is symlinks, eg:
  root@bsd:~# ln -s /usr/bin/lzop /usr/local/bin/lzop
```
It should be:
```
   ln -s /usr/local/bin/lzop  /usr/bin/lzop
```
Also needed
```
  ln -s /usr/local/bin/pv /usr/bin/pv
```
Moving all the FreeBSD notes to the FREEBSD.readme is cleaner -imho.

**Comment by jimsalterjrs** [comment]
If you want to submit a PR I'm happy to review it.

**Comment by jimsalterjrs** [comment]
Closing for now b/c I don't have time to spin up a FreeBSD VM and make sure the syntax is correct.

If any day-to-day FreeBSD / FreeNAS users would like to submit PRs regarding FreeBSD interoperability, I absolutely *will* accept PRs. =)

**Comment by phryneas** [comment]
Just as a thought: is there a reason to use hardcoded paths to the binaries at all? 
After all, this is what $PATH and /usr/bin/env are made for (as far as I know, the hard path /usr/bin/env is the only reliable existant path in almost every distribution/OS).

If used with cron, it should be encouraged anyways to write a correct PATH definition as a line at the beginning of the file or in front of the syncoid command. If invoked interactively, PATH should already be set up correctly.

TL;DR: Using /usr/bin/env to specify binary paths would be an easy way to get compatiblity with FreeBSD and even stuff like NixOs

**Comment by jimsalterjrs** [comment]
It would also immediately break the vast majority of existing installations due to lack of PATH set in cron, in one fell swoop.

Any thoughts about how to do this WITHOUT breaking existing installations - or making it more difficult for the majority of users to successfully automate the tasks (which this would, as things stand)?

**Comment by redmop** [comment]
source /etc/profile in the script itself?

**Comment by jimsalterjrs** [comment]
> source /etc/profile in the script itself?

... huh.

That's an interesting idea, but I'm feeling a little cautious about it.  It would break any PATH set by a user for their own interactive environment, correct?

Can anybody think of any reason we'd actually care about that?

Are there any platforms that don't use `/etc/profile`, for which this would break things in new and interesting ways?

Are there any existing examples of similar tools solving PATH-related woes this way?

**Comment by phryneas** [comment]
Well, for example current ubuntu's `/etc/crontab` defines PATH as
PATH=/usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin

So we have /usr/bin, we have /bin and we have /sbin - that makes all paths where syncoid has hardcoded paths to and some more.

I guess most other distributions handle this similarly, so I don't think using only PATH would really break that much. I might be wrong, but I think this might be something worth investigating.

To keep compatibility and break nobody's current setup, an idea could be to check in PATH and falling back to the current hard-coded paths if the commands are not found?

**Comment by jimsalterjrs** [comment]
yeah, things failing in normal entries in users' or root's cron is the big issue here.  For example I typically put sanoid and syncoid jobs in root's crontab.

> PPS: why not just add /sbin:/bin:/usr/bin at the end of the processes' current PATH?

Hm. Maybe. Do you know off the top of your head how you alter the current environment variables in Perl?  It's not something I've really mucked about with before.

**Comment by phryneas** [comment]
Sorry, no Perl experience whatsoever :(

If there's no way of directly setting it, my only guess would be that you can always do a call like `/usr/bin/env PATH="$PATH:/sbin:/bin:/usr/bin" zsh`

**Comment by stardude900** [comment]
So setting an environment variable in perl should only apply to the script being run.  The user's path shouldn't be affected.

Same applies to sourcing /etc/profile it'll only apply to the script, or at least that's how it is working on my Fedora 26 and FreeBSD 11.0 VMs right now.

```
#!/usr/bin/perl

# Should print a blank line, no path specified yet

system("/bin/echo $PATH");
$ENV{'PATH'} = '/bin'; 

# Should print /bin as that's the only thing specified in the path above.
print $ENV{'PATH'} ;
print ;
```

**Comment by phreaker0** [comment]
related to #21
